paddle/operators/elementwise_max_op.h

```diff
@@ -0,0 +1,120 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserve.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#pragma once
+
+#include "paddle/operators/elementwise_op_function.h"
+
+namespace paddle {
+namespace operators {
+
+template <typename T>
+struct MaxFunctor {
+  inline HOSTDEVICE T operator()(T a, T b) const { return a > b ? a : b; }
+};
+
+template <typename DeviceContext, typename T>
+class ElementwiseMaxKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& ctx) const override {
+    ElementwiseComputeEx<MaxFunctor<T>, DeviceContext, T>(ctx);
+  }
+};
+
+template <typename T>
+struct ElementwiseMaxGradFunctor {
+  template <typename Device, typename X, typename Y, typename Z, typename dX,
+            typename dY, typename dZ>
+  void operator()(Device d, X x, Y y, Z z, dX dx, dY dy, dZ dz) {
+    auto x_e = framework::EigenVector<T>::Flatten(*x);
+    auto y_e = framework::EigenVector<T>::Flatten(*y);
+    auto dz_e = framework::EigenVector<T>::Flatten(*dz);
+
+    if (dx) {
+      auto dx_e = framework::EigenVector<T>::Flatten(*dx);
+      dx_e.device(d) = (x_e > y_e).template cast<T>() * dz_e;
+    }
+    if (dy) {
+      auto dy_e = framework::EigenVector<T>::Flatten(*dy);
+      dy_e.device(d) = (x_e <= y_e).template cast<T>() * dz_e;
+    }
+  }
+};
+
```

Comment:
@JiayiFeng  We can reduce more codes. But we can do it later, if we have more element-wise operators to implement. Anyway, let's find a general pattern and clean code later.
```cpp
template <typename T>
struct MaxGradFunctor1 {
  template <typename Device, typename Dxe, typename Xe, typename Ye, typename Dze>
  void operator()(Device d, Dxe dx_e, Xe, x_e, Ye, y_e, Dze, dz_e) {
    dx_e.device(d) = (x_e > y_e).template cast<T>() * dz_e;
  }
};

template <typename T>
struct MinGradFunctor1 {
  template <typename Device, typename Dxe, typename Xe, typename Ye, typename Dze>
  void operator()(Device d, Dxe dx_e, Xe, x_e, Ye, y_e, Dze, dz_e) {
    dx_e.device(d) = (x_e < y_e).template cast<T>() * dz_e;
  }
};

template <typename T, typename Functor1, typename Functor2>
struct ElementwiseMinMaxGradFunctor {
  template <typename Device, typename X, typename Y, typename Z, typename dX,
            typename dY, typename dZ>
  void operator()(Device d, X x, Y y, Z z, dX dx, dY dy, dZ dz) {
    auto x_e = framework::EigenVector<T>::Flatten(*x);
    auto y_e = framework::EigenVector<T>::Flatten(*y);
    auto dz_e = framework::EigenVector<T>::Flatten(*dz);

    if (dx) {
      auto dx_e = framework::EigenVector<T>::Flatten(*dx);
      Functor1 functor;
      functor(d, dx_e, x_e, y_e, dz_e);
    }
    if (dy) {
      auto dy_e = framework::EigenVector<T>::Flatten(*dy);
      Functor2 functor;
      functor(d, dy_e, x_e, y_e, dz_e);
    }
  }
};
```

Reply:
@QiJune 
I agree. Find a general pattern in the backward pass is not as easy as that in forwarding pass, for backward `op` takes `Z`, `dZ`, `X`, `Y` while forwarding `op` only takes `X` and `Y`.